### PR TITLE
[Merged by Bors] - some cleanup for `bevy_ptr`

### PR DESCRIPTION
**Pull request by jakobhellermann**
1. change `PtrMut::as_ptr(self)` and `OwnedPtr::as_ptr(self)` to take `&self`, otherwise printing the pointer will prevent doing anything else afterwards
2. make all `as_ptr` methods safe. There's nothing unsafe about obtaining a pointer, these kinds of methods are safe in std as well [str::as_ptr](https://doc.rust-lang.org/stable/std/primitive.str.html#method.as_ptr), [Rc::as_ptr](https://doc.rust-lang.org/stable/std/rc/struct.Rc.html#method.as_ptr)
3. rename `offset`/`add` to `byte_offset`/`byte_add`. The unprefixed methods in std add in increments of `std::mem::size_of::<T>`, not in bytes. There's a PR for rust to add these byte_ methods https://github.com/rust-lang/rust/pull/95643 and at the call site it makes it much more clear that you need to do `.byte_add(i * layout_size)` instead of `.add(i)`

### CODE REVIEW

**Review by TheRawMeatball — APPROVED** [review]

**Review by BoxyUwU — APPROVED** [review]

File: crates/bevy_ptr/src/lib.rs [hunk 1]
```diff
@@ -116,13 +116,9 @@ impl<'a> Ptr<'a> {
...
     /// If possible, it is strongly encouraged to use [`deref`](Self::deref) over this function,
     /// as it retains the lifetime.
-    ///
-    /// # Safety
-    /// All subsequent operations to the returned pointer must be valid inside the
-    /// associated lifetime.
     #[inline]
     #[allow(clippy::wrong_self_convention)]
-    pub unsafe fn as_ptr(self) -> *mut u8 {
+    pub fn as_ptr(self) -> *mut u8 {
```
Comment (BoxyUwU) [inline crates/bevy_ptr/src/lib.rs:121]: this isnt taking `&self`

Comment (BoxyUwU) [reply crates/bevy_ptr/src/lib.rs:121]: it probably doesnt matter but idk if we want to be consistent here or not :sweat_smile:

**Review by james7132 — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Ping @cart for an `A-Pointer` tag.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**